xapian-letor/tests/api_letor.cc

```diff
@@ -0,0 +1,165 @@
+
+    return true;
+}
+
+static void
+populate_db(Xapian::WritableDatabase& db, const string&)
+{
+    Xapian::Document doc;
+    Xapian::TermGenerator termgenerator;
+    termgenerator.set_document(doc);
+    termgenerator.set_stemmer(Xapian::Stem("en"));
+    termgenerator.index_text("Lions, Tigers, Bears and Giraffes", 1, "S");
+    termgenerator.index_text("This paragraph talks about lions and tigers and "
+			     "bears (oh, my!). It mentions giraffes, "
+			     "but that's not really very important. Lions "
+			     "and tigers are big cats, so they must be really "
+			     "cuddly. Bears are famous for being cuddly, at "
+			     "least when they're teddy bears.", 1, "XD");
+    termgenerator.index_text("Lions, Tigers, Bears and Giraffes");
+    termgenerator.increase_termpos();
+    termgenerator.index_text("This paragraph talks about lions and tigers and "
+			     "bears (oh, my!). It mentions giraffes, "
+			     "but that's not really very important. Lions "
+			     "and tigers are big cats, so they must be really "
+			     "cuddly. Bears are famous for being cuddly, at "
+			     "least when they're teddy bears.");
+    db.add_document(doc);
+    doc.clear_terms();
+    termgenerator.index_text("Lions, Tigers and Bears", 1, "S");
+    termgenerator.index_text("This is the paragraph of interest. Tigers are "
+			     "massive beasts - I wouldn't want to meet a "
+			     "hungry one anywhere. Lions are scary even when "
+			     "lyin' down. Bears are scary even when bare. "
+			     "Together I suspect they'd be less scary, as the "
+			     "tigers, lions, and bears would all keep each "
+			     "other busy. On the other hand, bears don't live "
+			     "in the same continent as far as I know.", 1,
+			     "XD");
+    termgenerator.index_text("Lions, Tigers and Bears");
+    termgenerator.increase_termpos();
+    termgenerator.index_text("This is the paragraph of interest. Tigers are "
+			     "massive beasts - I wouldn't want to meet a "
+			     "hungry one anywhere. Lions are scary even when "
+			     "lyin' down. Bears are scary even when bare. "
+			     "Together I suspect they'd be less scary, as the "
+			     "tigers, lions, and bears would all keep each "
+			     "other busy. On the other hand, bears don't live "
+			     "in the same continent as far as I know.");
+    db.add_document(doc);
+}
+
+DEFINE_TESTCASE(listnet_ranker, generated) {
+    Xapian::ListNETRanker ranker;
+    string a = get_database_path("apitest_ranker1",
+				       populate_db);
+    Xapian::Enquire enquire((Xapian::Database(a)));
+    enquire.set_query(Xapian::Query("lions"));
+    Xapian::MSet mymset = enquire.get_mset(0, 10);
+    Xapian::prepare_training_file(a, "testdata/query.txt",
+				  "testdata/qrel.txt", 10,
+				  "training-data.txt");
+    ranker.set_database_path(a);
+    ranker.set_query(Xapian::Query("lions"));
+    ranker.train_model("training-data.txt", "");
```

Comment:
I can read the API docs to see that - but why did you choose to pass `""` when you wrote this testcase?

There seem to really be three cases here - there's specifying a filename and a `model_key`, there's just specifying a filename, and there's specifying a filename and an empty string for `model_key`.

Of these three, I'd imagine that the first and second are the ones API users are most likely to use, with probably second being more common.  The main case I can see where the third case is useful is when the API user's code makes `model_key` configurable to the end-user.

You're only testing one of these three cases here, and it seems to me it is the case which is probably least important.